src/PDistributions/PDistributions.jl

```diff
@@ -0,0 +1,377 @@
+"""
+  particle mass distribution module
+
+Particle mass distribution functions for microphysical process modeling:
+  - computation of moments of distributions
+  - computation of densities of distributions
+  - creating distributions given a set of parameters
+  - creating distributions given a set of moments
+"""
+module PDistributions
+
+using SpecialFunctions: gamma, gamma_inc
+using DocStringExtensions
+
+import LinearAlgebra: norm
+import Optim: optimize, LBFGS
+
+# mass distributions available for microphysics
+export PDistribution
+export Primitive
+export Exponential
+export Gamma
+export Mixture
+
+# methods that query particle mass distributions
+export moment
+export density
+
+# setters and getters
+export update_params_from_moments
+
+
+"""
+  PDistribution{FT}
+
+A particle mass distribution function, which can be initialized
+for various subtypes of assumed shapes in the microphysics parameterization.
+"""
+abstract type PDistribution{FT} end
+
+
+"""
+  Primitive{FT}
+
+A particle mass distribution that has support on the positive real
+axis and analytic expressions for moments and partial moments.
+"""
+abstract type Primitive{FT} <: PDistribution{FT} end
+
+
+"""
+  Exponential{FT} <: Primitive{FT}
+
+Represents particle mass distribution function of exponential shape.
+
+# Constructors
+  Exponential(n::Real, θ::Real)
+
+# Fields
+$(DocStringExtensions.FIELDS)
+"""
+struct Exponential{FT} <: Primitive{FT}
+  "normalization constant (e.g., droplet number concentration)"
+  n::FT
+  "scale parameter"
+  θ::FT
+
+  function Exponential(n::FT, θ::FT) where {FT<:Real}
+    if n < 0 || θ <= 0
+      error("n needs to be nonnegative. θ needs to be positive.")
+    end
+
+    new{FT}(n, θ)
+  end
+end
+
+
+"""
+  Gamma{FT} <: Primitive{FT}
+
+Represents particle mass distribution function of gamma shape.
+
+# Constructors
+  Gamma(n::Real, θ::Real, k::Real)
+
+# Fields
+$(DocStringExtensions.FIELDS)
+"""
+struct Gamma{FT} <: Primitive{FT}
+  "normalization constant (e.g., droplet number concentration)"
+  n::FT
+  "scale parameter"
+  θ::FT
+  "shape parameter"
+  k::FT
+
+  function Gamma(n::FT, θ::FT, k::FT) where {FT<:Real}
+    if n < 0 || θ <= 0 || k <= 0
+      error("n needs to be nonnegative. θ and k need to be positive.")
+    end
+    new{FT}(n, θ, k)
+  end
+end
+
+
+"""
+  Mixture{FT} <: PDistribution{FT}
+
+A particle mass distribution function that is a mixture of
+primitive or truncated subdistribution functions.
+
+# Constructors
+  Mixture(dists::PDistribution{Real}...)
+  Mixture(dist_arr::Array{PDistribution{FT}})
+
+# Fields
+$(DocStringExtensions.FIELDS)
+"""
+struct Mixture{FT} <: PDistribution{FT}
+  "array of distributions"
+  subdists::Array{PDistribution{FT}}
```

Comment:
Do we know the rank of `subdists`? Adding the rank may be useful for the compiler.